delivery/common/jwt/lib/jwt.js

```diff
@@ -0,0 +1,132 @@
+import { crypto } from "crypto";
+
+import { base64url, TextEncoder } from "encoding";
+
+class JWTUtil {
+    static isEmptyString(str) {
+        return !str || 0 === str.trim().length;
+    }
+}
+
+class JWTValidator {
+    constructor(jwtOptions) {
+        this.jwtOptions = jwtOptions || {}, this.validateOptionTypes(), this.algorithms = [ "RS256", "RS384", "RS512", "HS256", "HS384", "HS512", "PS256", "PS384", "PS512", "ES256", "ES384", "ES512" ], 
+        this.jwtOptions.allowUnsecuredToken && this.algorithms.push("NONE");
+    }
+    async validate(base64JWTToken, keys) {
+        var _a, _b;
+        if ("string" != typeof base64JWTToken) throw new Error("Invalid token type, expected string!");
+        if (!Array.isArray(keys) || !keys.every((elem => void 0 !== elem.type || void 0 !== elem.extractable || null != elem.algorithm || null != elem.usages))) throw new Error("Invalid keys type, expected list of CryptoKey!");
+        const jwtParts = base64JWTToken.split(".");
+        if (jwtParts.length > 3 || jwtParts.length < 2) throw new Error("JWT malformed: Invalid number of parts for JWT token. expected 3 or 2 (unsecured JWT)!");
+        if (JWTUtil.isEmptyString(jwtParts[0])) throw new Error("JWT malformed: jwt header cannot be empty");
+        if (JWTUtil.isEmptyString(jwtParts[1])) throw new Error("JWT malformed: jwt payload cannot be empty");
+        const jwtHBin = base64url.decode(jwtParts[0], "String"), jwtPBin = base64url.decode(jwtParts[1], "String"), jwtHeader = JSON.parse(jwtHBin), jwtPayload = JSON.parse(jwtPBin);
+        if (this.jwtOptions.issuer && jwtPayload.iss && this.jwtOptions.issuer !== jwtPayload.iss) throw new Error(`JWT malformed: invalid iss, expected ${this.jwtOptions.issuer}`);
+        if (this.jwtOptions.subject && jwtPayload.sub && this.jwtOptions.subject !== jwtPayload.sub) throw new Error(`JWT malformed: invalid sub, expected ${this.jwtOptions.subject}`);
+        if (this.jwtOptions.audience) {
+            const aud = jwtPayload.aud;
+            if (aud && (Array.isArray(aud) || "string" == typeof aud)) {
+                if (!(Array.isArray(aud) ? aud : [ aud ]).includes(this.jwtOptions.audience)) throw new Error(`JWT malformed: invalid aud, expected ${this.jwtOptions.audience}`);
```

Comment:
This is just a style comment, feel free to ignore it: 

Is it worth distinguishing between malformed JWTs and JWTs that fail to validate?

Reply:
As of now we return JWT malformed when the type check fails for any JWT fields as per the JWT spec. e.g is JWT header  or payload is empty. However for cases like arguments validation failure or signature validation failures we don't return error indicating JWT malformed. 
May be we can keep as it is for now.